python/ray/dag/compiled_dag_node.py

```diff
@@ -0,0 +1,331 @@
+from typing import Any, List, Tuple, Union
+
+import ray
+import ray.experimental.channel as ray_channel
+
+
+MAX_BUFFER_SIZE = int(100 * 1e6)  # 100MB
+
+ChannelType = "ray.experimental.channel.Channel"
+
+
+def allocate_channel(buffer_size_bytes: int = MAX_BUFFER_SIZE, num_readers: int = 1):
+    if not isinstance(buffer_size_bytes, int):
+        raise ValueError("buffer_size_bytes must be an integer")
+    if not isinstance(num_readers, int):
+        raise ValueError("num_readers must be an integer")
+
+    return ray_channel.Channel(buffer_size_bytes, num_readers)
+
+
+def do_allocate_channel(
+    self, buffer_size_bytes: int = MAX_BUFFER_SIZE, num_readers: int = 1
+):
+    self._output_channel = allocate_channel(buffer_size_bytes)
+    return self._output_channel
+
+
+def do_exec_compiled_task(
+    self,
+    inputs: List[Union[Any, "ray_channel.Channel"]],
+    actor_method_name: str,
+):
+    try:
+        method = getattr(self, actor_method_name)
+
+        resolved_inputs = []
+        input_channel_idxs = []
+        # Add placeholders for input channels.
+        for inp in inputs:
+            if isinstance(inp, ray_channel.Channel):
+                input_channel_idxs.append((len(resolved_inputs), inp))
+                resolved_inputs.append(None)
+            else:
+                resolved_inputs.append(inp)
+
+        while True:
+            for idx, chan in input_channel_idxs:
+                resolved_inputs[idx] = chan.begin_read()
+
+            output_val = method(*resolved_inputs)
+
+            self._output_channel.write(output_val)
+            for _, chan in input_channel_idxs:
+                chan.end_read()
+
+    except Exception as e:
+        print("Task aborted", e)
+        raise
+
+
+class CompiledTask:
+    """Wraps the normal Ray DAGNode with some metadata."""
+
+    def __init__(self, idx, dag_node: "ray.dag.DAGNode"):
+        self.idx = idx
+        self.dag_node = dag_node
+
+        self.args = []
+        self.dependent_node_idxs = set()
+        self.output_channel = None
+
+    @property
+    def num_readers(self):
+        return len(self.dependent_node_idxs)
+
+    def __str__(self):
+        return f"""
+Node: {self.dag_node}
+Arguments: {self.args}
+Output: {self.output_channel}
+"""
+
+
+class CompiledDAG:
```

Comment:
Should we add tests that validate DAG itself (without running execute)? 

1. Convert regular DAG and make sure returned compiled dag contains all regular DAG nodes + correct worker_task_refs
2. Test cases where exceptions are raised